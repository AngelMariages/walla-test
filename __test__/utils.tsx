import { render } from '@testing-library/react';
import {
    FavoritesContext,
    FavoritesContextType,
} from 'context/FavoritesContext';
import { FiltersContext, FiltersContextType } from 'context/FiltersContext';

type ContextValues = {
    filtersContext?: Partial<FiltersContextType>;
    favoritesContext?: Partial<FavoritesContextType>;
};

export const renderWithContext = (
    ui: React.ReactElement,
    contextValues?: ContextValues
) => {
    return render(
        <FiltersContext.Provider
            value={{
                search: '',
                setSearch: jest.fn(),
                page: 1,
                incrementPage: jest.fn(),
                sort: {
                    sortBy: 'title',
                    sortOrder: 'asc',
                },
                setSort: jest.fn(),
                ...(contextValues?.filtersContext || {}),
            }}>
            <FavoritesContext.Provider
                value={{
                    favorites: [],
                    favoritesVisible: false,
                    addFavorite: jest.fn(),
                    removeFavorite: jest.fn(),
                    isFavorite: jest.fn(),
                    toggleFavorites: jest.fn(),
                    ...(contextValues?.favoritesContext || {}),
                }}>
                {ui}
            </FavoritesContext.Provider>
        </FiltersContext.Provider>
    );
};
