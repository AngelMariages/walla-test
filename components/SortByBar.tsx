import { useContext } from 'react';
import { Sort, sortBys } from '../constants/sortBy';
import { FiltersContext } from '../context/FiltersContext';
import UpArrow from 'public/icons/arrow-up.svg';
import DownArrow from 'public/icons/arrow-down.svg';

type SortArrowProps = {
    sortOrder: Sort['sortOrder'];
    className?: string;
};

const SortArrow: React.FC<SortArrowProps> = ({ sortOrder, className }) => {
    if (sortOrder === 'asc') {
        return <UpArrow className={className} />;
    } else {
        return <DownArrow className={className} />;
    }
};

const capitalizeFirstLetter = (string: string) => {
    return string.charAt(0).toUpperCase() + string.slice(1);
};

const SortByBar = () => {
    const { setSort, sort } = useContext(FiltersContext);

    return (
        <div className="flex flex-row py-4 px-4 justify-center">
            <div className="flex flex-row pl-4 gap-4 text-white">
                {sortBys.map((sortBy) => (
                    <button
                        key={sortBy}
                        className="px-3 md:px-4 rounded-lg bg-[#12C1AC] flex flex-row items-center"
                        onClick={() => {
                            setSort(sortBy);
                        }}>
                        {capitalizeFirstLetter(sortBy)}
                        {sort.sortBy === sortBy && (
                            <SortArrow
                                sortOrder={sort.sortOrder}
                                className="ml-2 w-4 h-4"
                            />
                        )}
                    </button>
                ))}
            </div>
        </div>
    );
};

export default SortByBar;
