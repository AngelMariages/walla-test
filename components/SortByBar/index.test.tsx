import { fireEvent, getByTestId } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { renderWithContext } from '__test__/utils';
import SortByBar from '.';

describe('SortByBar', () => {
    it('should render correctly', () => {
        const { container } = renderWithContext(<SortByBar />);
        expect(container).toMatchSnapshot();
    });

    it('should render correctly with asc', () => {
        const { container } = renderWithContext(<SortByBar />, {
            filtersContext: {
                sort: {
                    sortBy: 'title',
                    sortOrder: 'asc',
                },
            },
        });
        expect(container).toMatchSnapshot();

        const ascButton = getByTestId(container, 'sort-by-title-asc');

        expect(ascButton).toBeDefined();
    });

    it('should call setSort when button is clicked', async () => {
        const setSort = jest.fn();
        const { getByTestId } = renderWithContext(<SortByBar />, {
            filtersContext: {
                setSort,
            },
        });
        const button = getByTestId('sort-by-title-asc');

        await act(() => {
            fireEvent.click(button);
        });
        expect(setSort).toHaveBeenCalledTimes(1);
        expect(setSort).toHaveBeenCalledWith('title');

        const button2 = getByTestId('sort-by-price-asc');

        await act(() => {
            fireEvent.click(button2);
        });
        expect(setSort).toHaveBeenCalledTimes(2);
        expect(setSort).toHaveBeenCalledWith('price');
    });
});
