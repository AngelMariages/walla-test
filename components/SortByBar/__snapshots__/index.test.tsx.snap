// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortByBar should render correctly 1`] = `
<div>
  <div
    class="flex flex-row py-4 px-4 justify-center"
  >
    <div
      class="flex flex-row pl-4 gap-4 text-white"
    >
      <button
        class="px-3 md:px-4 rounded-lg bg-[#12C1AC] flex flex-row items-center"
        data-testid="sort-by-title-asc"
      >
        Title
        <svg-url
          classname="ml-2 w-4 h-4"
        />
      </button>
      <button
        class="px-3 md:px-4 rounded-lg bg-[#12C1AC] flex flex-row items-center"
        data-testid="sort-by-description-asc"
      >
        Description
      </button>
      <button
        class="px-3 md:px-4 rounded-lg bg-[#12C1AC] flex flex-row items-center"
        data-testid="sort-by-price-asc"
      >
        Price
      </button>
      <button
        class="px-3 md:px-4 rounded-lg bg-[#12C1AC] flex flex-row items-center"
        data-testid="sort-by-email-asc"
      >
        Email
      </button>
    </div>
  </div>
</div>
`;

exports[`SortByBar should render correctly with asc 1`] = `
<div>
  <div
    class="flex flex-row py-4 px-4 justify-center"
  >
    <div
      class="flex flex-row pl-4 gap-4 text-white"
    >
      <button
        class="px-3 md:px-4 rounded-lg bg-[#12C1AC] flex flex-row items-center"
        data-testid="sort-by-title-asc"
      >
        Title
        <svg-url
          classname="ml-2 w-4 h-4"
        />
      </button>
      <button
        class="px-3 md:px-4 rounded-lg bg-[#12C1AC] flex flex-row items-center"
        data-testid="sort-by-description-asc"
      >
        Description
      </button>
      <button
        class="px-3 md:px-4 rounded-lg bg-[#12C1AC] flex flex-row items-center"
        data-testid="sort-by-price-asc"
      >
        Price
      </button>
      <button
        class="px-3 md:px-4 rounded-lg bg-[#12C1AC] flex flex-row items-center"
        data-testid="sort-by-email-asc"
      >
        Email
      </button>
    </div>
  </div>
</div>
`;
